src/core/test/system/borrower/borrow.t.sol

```diff
@@ -0,0 +1,151 @@
+// Copyright (C) 2019 Centrifuge
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.5.12;
+
+import "../system.sol";
+import "../users/borrower.sol";
+import "../users/admin.sol";
+
+contract BorrowTest is SystemTest {
+
+    Borrower borrower;
+    address borrower_;
+
+    Borrower randomUser;
+    address randomUser_;
+    
+    AdminUser public admin;
+    address admin_;
+        
+    function setUp() public {
+        baseSetup();
+        // setup users
+        borrower = new Borrower(address(shelf), address(distributor), currency_, address(pile));
+        borrower_ = address(borrower);
+        randomUser = new Borrower(address(shelf), address(distributor), currency_, address(pile));
+        randomUser_ = address(randomUser);
+        admin = new AdminUser(address(shelf), address(pile), address(ceiling), address(title));
+        admin_ = address(admin);
+
+        rootAdmin.relyBorrowAdmin(admin_);
+    }
+    
+    function borrow(uint loanId, uint tokenId, uint amount) public {
+        uint initialTotalBalance = shelf.balance();
+        uint initialLoanBalance = shelf.balances(loanId);
+        uint initialTotalDebt = pile.total();
+        uint initialLoanDebt = pile.debt(loanId);
+        uint initialCeiling = ceiling.ceiling(loanId);
+
+        borrower.borrow(loanId, amount);
+        assertPostCondition(loanId, tokenId, amount, initialTotalBalance, initialLoanBalance, initialTotalDebt, initialLoanDebt, initialCeiling);
+    }
+
+    function lockNFT(uint loanId) public {
+        borrower.approveNFT(collateralNFT, address(shelf));
+        borrower.lock(loanId);
+    }
+
+    function assertPreCondition(uint loanId, uint tokenId, uint amount) public {
+        // assert: borrower loanOwner
+        assertEq(title.ownerOf(loanId), borrower_);
+        // assert: shelf nftOwner
+        assertEq(collateralNFT.ownerOf(tokenId), address(shelf));
+        // assert: borrowAmount <= ceiling
+        assert(amount <= ceiling.ceiling(loanId));
```

Comment:
It should be the following
```suggestion
        assert(amount < ceiling.ceiling(loanId));
```
 because the require is  `require(values[loan] >= amount);` or we change the require in the ceilings

Reply:
I think we should change it in the principal and creditline to:

 `require(add(pile.debt(loan), amount) <= values[loan]);`

Reply:
we discussed with @lucasvo and agreed to consider debt just for the creditLine, not for principal